### ISSUE TITLE
lack of NetworkDriver.ProgramExternalConnectivity method causes failure at docker 1.12

### ISSUE THREAD

**Issue by ekesken** [projectcalico/libnetwork-plugin#60]
I use calico-libnetwork v0.8.0, these are relevant logs:

```
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1476, in _find_error_handler
    .get(code))
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1465, in find_handler
    handler = handler_map.get(cls)
AttributeError: 'function' object has no attribute 'get'
 - - [12/Aug/2016:11:46:39 +0000] "POST /NetworkDriver.ProgramExternalConnectivity HTTP/1.1" 500 - "-" "-"
```

I checked repo and saw following two methods are not implemted yet:

```
/NetworkDriver.ProgramExternalConnectivity
/NetworkDriver.RevokeExternalConnectivity
```

**Comment by ekesken** [comment]
it seems it does not work either in docker 1.11, this is the change breaks compatibility:

https://github.com/docker/docker/commit/5a654089416762da484ca9929fcd216076b591ab

these are the full logs with docker 1.11.2:

calico-libnetwork:
```
[2016-08-12 13:24:33 +0000] [25069] [INFO] Starting gunicorn 19.6.0
[2016-08-12 13:24:33 +0000] [25069] [INFO] Listening at: unix:/run/docker/plugins/calico.sock (25069)
[2016-08-12 13:24:33 +0000] [25069] [INFO] Using worker: gevent
[2016-08-12 13:24:33 +0000] [25075] [INFO] Booting worker with pid: 25075
Application started
Activate response JSON={'Implements': ['NetworkDriver', 'IpamDriver']}
 - - [12/Aug/2016:13:30:34 +0000] "POST /Plugin.Activate HTTP/1.1" 200 65 "-" "Go-http-client/1.1"
GetCapabilities response JSON={'Scope': 'global'}
 - - [12/Aug/2016:13:30:34 +0000] "POST /NetworkDriver.GetCapabilities HTTP/1.1" 200 24 "-" "Go-http-client/1.1"
[2016-08-12 13:30:34 +0000] [25075] [ERROR] Error handling request /IpamDriver.GetCapabilities
Traceback (most recent call last):
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 52, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/ggevent.py", line 152, in handle_request
    super(GeventWorker, self).handle_request(*args)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 103, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 2000, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1991, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1559, in handle_exception
    handler = self._find_error_handler(InternalServerError())
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1476, in _find_error_handler
    .get(code))
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1465, in find_handler
    handler = handler_map.get(cls)
AttributeError: 'function' object has no attribute 'get'
 - - [12/Aug/2016:13:30:34 +0000] "POST /IpamDriver.GetCapabilities HTTP/1.1" 500 - "-" "-"
GetDefaultAddressSpace response JSON={'GlobalDefaultAddressSpace': 'CalicoGlobalAddressSpace', 'LocalDefaultAddressSpace': 'CalicoLocalAddressSpace'}
RequestAddress JSON={u'Address': u'172.28.254.200', u'Options': None, u'PoolID': u'172.28.0.0/16'}
Reserving a specific address in Calico pools
RequestAddress response JSON={'Data': {}, 'Address': '172.28.254.200/32'}
 - - [12/Aug/2016:13:30:34 +0000] "POST /IpamDriver.RequestAddress HTTP/1.1" 200 52 "-" "Go-http-client/1.1"
CreateEndpoint JSON={u'NetworkID': u'30baf2d6317bb29ba6f304842b945b9fac365fb03998596f80d20be0e9b05e40', u'Interface': {u'AddressIPv6': u'', u'MacAddress': u'', u'Address': u'172.28.254.200/32'}, u'Options': {u'com.docker.network.endpoint.exposedports': [], u'com.docker.network.portmap': []}, u'EndpointID': u'7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc'}
Creating endpoint 7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc
No IPv6 data
Using Calico IPAM driver, get next hops
Saving Calico endpoint: Endpoint({"ipv6_gateway": null, "state": "active", "name": "cali7e4f3bb4c9d", "ipv4_gateway": "10.135.20.243", "ipv6_nets": [], "profile_ids": ["30baf2d6317bb29ba6f304842b945b9fac365fb03998596f80d20be0e9b05e40"], "mac": "EE:EE:EE:EE:EE:EE", "ipv4_nets": ["172.28.254.200/32"]})
CreateEndpoint response JSON={'Interface': {'MacAddress': 'EE:EE:EE:EE:EE:EE'}}
 - - [12/Aug/2016:13:30:34 +0000] "POST /NetworkDriver.CreateEndpoint HTTP/1.1" 200 63 "-" "Go-http-client/1.1"
Join JSON={u'NetworkID': u'30baf2d6317bb29ba6f304842b945b9fac365fb03998596f80d20be0e9b05e40', u'SandboxKey': u'/var/run/docker/netns/f082e047ae36', u'Options': {u'com.docker.network.endpoint.exposedports': [], u'com.docker.network.portmap': []}, u'EndpointID': u'7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc'}
Joining endpoint 7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc
No IPv6 data
Using Calico IPAM driver, configure gateway and static routes to the host
Join Response JSON={'StaticRoutes': [{'NextHop': '', 'Destination': '10.135.20.243/32', 'RouteType': 1}], 'InterfaceName': {'SrcName': u'tmp7e4f3bb4c9d', 'DstPrefix': 'cali'}, 'Gateway': '10.135.20.243'}
 - - [12/Aug/2016:13:30:34 +0000] "POST /NetworkDriver.Join HTTP/1.1" 200 241 "-" "Go-http-client/1.1"
[2016-08-12 13:30:34 +0000] [25075] [ERROR] Error handling request /NetworkDriver.ProgramExternalConnectivity
Traceback (most recent call last):
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 52, in handle
    self.handle_request(listener_name, req, client, addr)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/ggevent.py", line 152, in handle_request
    super(GeventWorker, self).handle_request(*args)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/gunicorn/workers/async.py", line 103, in handle_request
    respiter = self.wsgi(environ, resp.start_response)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 2000, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1991, in wsgi_app
    response = self.make_response(self.handle_exception(e))
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1559, in handle_exception
    handler = self._find_error_handler(InternalServerError())
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1476, in _find_error_handler
    .get(code))
  File "/usr/local/libnetwork-plugin/local/lib/python2.7/site-packages/flask/app.py", line 1465, in find_handler
    handler = handler_map.get(cls)
AttributeError: 'function' object has no attribute 'get'
 - - [12/Aug/2016:13:30:34 +0000] "POST /NetworkDriver.ProgramExternalConnectivity HTTP/1.1" 500 - "-" "-"
Leave JSON={u'NetworkID': u'30baf2d6317bb29ba6f304842b945b9fac365fb03998596f80d20be0e9b05e40', u'EndpointID': u'7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc'}
Leaving endpoint 7e4f3bb4c9dc3161c9f4452cd7ed4ba3eabdae51505fb0e3310474da585c5edc
Leave response JSON={}
 - - [12/Aug/2016:13:30:34 +0000] "POST /NetworkDriver.Leave HTTP/1.1" 200 3 "-" "Go-http-client/1.1"
```

docker:
```
time="2016-08-12T13:31:18.379020823Z" level=debug msg="Assigning addresses for endpoint zookeeper-teasers-tr-master-001's interface on network calico-net1"
time="2016-08-12T13:31:18.462101632Z" level=debug msg="Assigning addresses for endpoint zookeeper-teasers-tr-master-001's interface on network calico-net1"
time="2016-08-12T13:31:18.652545436Z" level=debug msg="Programming external connectivity on endpoint zookeeper-teasers-tr-master-001 (d95849edd24ed7c5956c54109dca115548ef7256d6f4bc85f8261baa6967942c)"
time="2016-08-12T13:31:18.732433377Z" level=warning msg="Could not rollback container connection to network calico-net1"
time="2016-08-12T13:31:18.762474247Z" level=warning msg="failed to cleanup ipc mounts:\nfailed to umount /var/lib/docker/containers/490bbe8bbd1a727af003227f85560e719e052a80273993747c5fdb62e216fa1e/shm: no such file or directory"
time="2016-08-12T13:31:18.793518538Z" level=error msg="Handler for POST /v1.23/containers/490bbe8bbd1a727af003227f85560e719e052a80273993747c5fdb62e216fa1e/start returned error: driver failed programming external connectivity on endpoint zookeeper-teasers-tr-master-001 (d95849edd24ed7c5956c54109dca115548ef7256d6f4bc85f8261baa6967942c): NetworkDriver.ProgramExternalConnectivity: <html>\n  <head>\n    <title>Internal Server Error</title>\n  </head>\n  <body>\n    <h1><p>Internal Server Error</p></h1>\n    \n  </body>\n</html>\n"
time="2016-08-12T13:31:18.793687555Z" level=debug msg="attach: stderr: end"
time="2016-08-12T13:31:18.793746345Z" level=debug msg="attach: stdout: end"
time="2016-08-12T13:31:18.795336748Z" level=debug msg="Calling DELETE /v1.23/containers/490bbe8bbd1a727af003227f85560e719e052a80273993747c5fdb62e216fa1e?v=1"
time="2016-08-12T13:31:18.798022703Z" level=debug msg="Removing volume reference: driver local, name f8d387de272ab85deb4a6720f1f4521c58958567c91dda61d268bcb4afe93979"
time="2016-08-12T13:31:18.834564937Z" level=debug msg="Calling DELETE /v1.23/containers/zookeeper-teasers-tr-master-001?force=1"
time="2016-08-12T13:31:18.834683604Z" level=error msg="Handler for DELETE /v1.23/containers/zookeeper-teasers-tr-master-001 returned error: No such container: zookeeper-teasers-tr-master-001"
time="2016-08-12T13:31:18.847038529Z" level=debug msg="Calling POST /v1.23/containers/create?name=zookeeper-teasers-tr-master-001"
```

and retrying "docker run" after, causes this error:

```
time="2016-08-12T13:31:19.071742669Z" level=error msg="Handler for POST /v1.23/containers/3de47fc0c68d3370c72a71ea5f6cbd3cf22dfb42e3485456a6bc0143cfc41678/start returned error: service endpoint with name zookeeper-teasers-tr-master-001 already exists"
```

**Comment by robbrockbank** [comment]
Hi @ekesken - Could you paste in the series of commands that cause the failure in the driver?  Thanks.

**Comment by ekesken** [comment]
Hey @robbrockbank - this is the docker run command:

```
/usr/bin/docker run --rm \
  --volume=/docker/zookeeper:/var/lib/zookeeper:rw \
  --net=calico-net1 \
  --ip=172.28.254.207 \
  --pid=host \
  --name=zookeeper-teasers-tr-master-003 \
  --env=ZOOKEEPER_ID="3" \
  --env=ZOOKEEPER_SERVER_1="172.28.254.205" \
  --env=ZOOKEEPER_SERVER_2="172.28.254.206" \
  --env=ZOOKEEPER_SERVER_3="172.28.254.207" \
  --env=ZOOKEEPER_SERVER_4="172.28.254.208" \
  --env=ZOOKEEPER_SERVER_5="172.28.254.209" \
  --env=ZOOKEEPER_AUTOPURGE_SNAP_RETAIN_COUNT=4 \
  --env=ZOOKEEPER_AUTOPURGE_PURGE_INTERVAL=6 \
  digitalwonderland/zookeeper \
  /usr/local/bin/start.sh
```

and these are the versions we use in our cluster:

```
ubuntu-16-04-x64
docker-engine=1.11.2-0~xenial or docker-engine=1.12.0-0~xenial
etcd-v2.3.7
calico/node:v0.20.0
https://github.com/projectcalico/libnetwork-plugin/tree/v0.8.0
```

**Comment by matthewdupre** [comment]
Let me quickly try a repro - @ekesken can you please show me the `docker network create` command you used?

**Comment by ekesken** [comment]
hey @matthewdupre, this is the command:

```
    docker network create \
      --driver calico \
      --ipam-driver calico \
      --subnet=172.28.0.0/16 \
      calico-net1
```

**Comment by matthewdupre** [comment]
@ekesken I just tried this out on a fresh 16.04, and didn't see any problems.

Let me give you a few details of my install / setup process, and you can point out where there are any differences:
- fresh 16.04 VM on GCE
- added the repo and installed docker-engine according to https://docs.docker.com/engine/installation/linux/ubuntulinux/.  This gave me `Docker version 1.12.0, build 8eab29e`.
- installed and started etcd
- added `--cluster-store=etcd://localhost:2379` to docker daemon and restarted docker
- downloaded calicoctl 0.20, `calicoctl node --libnetwork`
- docker network create as above
- docker run as above

and it worked (well, actually zookeeper complained about creating /var/lib/zookeeper, but all the networking succeeded - I also checked pings against some busyboxes).  I also checked docker logs (journalctl -u docker); there was nothing out of the ordinary.

How does your setup differ from this?

**Comment by ekesken** [comment]
@matthewdupre I created a gist to let you reproduce the problem easily:

https://gist.github.com/ekesken/b193efba45676cdbdf6f79e8e0ad627b 

try following commands:

```
git clone https://gist.github.com/ekesken/b193efba45676cdbdf6f79e8e0ad627b libnetwork_plugin_issue_60
cd libnetwork_plugin_issue_60
docker build -t libnetwork_plugin_issue_60 .
docker run --privileged --rm libnetwork_plugin_issue_60
```

and I confirm that problem is solved by v0.9.0 release published 4 days ago, libnetwork plugin returns 404 to ProgramExternalConnectivity requests instead of 500 in new release.

you can see it's solved via following commands:

```
sed -i 's/v0.8.0/v0.9.0/g' Dockerfile`
docker build -t libnetwork_plugin_issue_60 .
docker run --privileged --rm libnetwork_plugin_issue_60
```

**Comment by matthewdupre** [comment]
Fixed in 0.9: closing.  Thanks ekesken :)!
